powerauth-restful-security-spring/src/main/java/io/getlime/security/powerauth/rest/api/spring/annotation/PowerAuthAnnotationInterceptor.java

```diff
@@ -96,14 +120,103 @@ public boolean preHandle(HttpServletRequest request, HttpServletResponse respons
     }
 
+    /**
+     * Decrypt HTTP request body and construct object with ECIES data.
+     *
+     * @param request HTTP request.
+     * @return Object with ECIES data.
+     * @throws PowerAuthEncryptionException In case request decryption fails.
+     */
+    private PowerAuthEciesEncryption decryptRequest(HttpServletRequest request) throws PowerAuthEncryptionException {
+        // Read ECIES metadata from HTTP header
+        final PowerAuthEciesEncryption eciesEncryption = this.encryptionProvider.prepareEciesEncryption(request.getHeader(PowerAuthEncryptionHttpHeader.HEADER_NAME));
+
+        try {
+            // Extract request body from HTTP request
+            Scanner scanner = new Scanner(request.getInputStream()).useDelimiter("\\A");
+            String requestBody = scanner.hasNext() ? scanner.next() : null;
+            if (requestBody == null) {
+                throw new PowerAuthEncryptionException("Invalid HTTP request");
+            }
+
+            // Parse ECIES cryptogram from request body
+            final EciesCryptogram eciesCryptogram = parseEciesCryptogram(requestBody);
+
+            // Prepare ephemeral public key
+            final String ephemeralPublicKey = BaseEncoding.base64().encode(eciesCryptogram.getEphemeralPublicKey());
+
+            // Obtain ECIES decryptor parameters from PowerAuth server
+            final PowerAuthEciesDecryptorParameters decryptorParameters = this.encryptionProvider.getEciesDecryptorParameters(eciesEncryption.getActivationId(),
+                    eciesEncryption.getApplicationKey(), ephemeralPublicKey);
+
+            // Prepare envelope key and sharedInfo2 parameter for decryptor
+            final byte[] secretKey = BaseEncoding.base64().decode(decryptorParameters.getSecretKey());
+            final EciesEnvelopeKey envelopeKey = new EciesEnvelopeKey(secretKey, eciesCryptogram.getEphemeralPublicKey());
+            final byte[] sharedInfo2 = BaseEncoding.base64().decode(decryptorParameters.getSharedInfo2());
+
+            // Construct decryptor and set it to the request for later encryption of response
+            final EciesDecryptor eciesDecryptor = eciesFactory.getEciesDecryptor(envelopeKey, sharedInfo2);
+            eciesEncryption.setEciesDecryptor(eciesDecryptor);
+
+            // Decrypt request data
+            byte[] decryptedData = eciesDecryptor.decryptRequest(eciesCryptogram);
+            eciesEncryption.setDecryptedRequest(decryptedData);
+        } catch (Exception ex) {
+            throw new PowerAuthEncryptionException("Invalid HTTP request");
+        }
+
+        return eciesEncryption;
+    }
+
+    /**
+     * Read ECIES cryptogram from HTTP request.
+     *
+     * @param requestBody Request data in JSON format.
+     * @return ECIES Cryptogram.
+     * @throws PowerAuthEncryptionException In case JSON parsing fails.
+     */
+    private EciesCryptogram parseEciesCryptogram(String requestBody) throws PowerAuthEncryptionException {
```

Comment:
The parsing code below is unnecessary, the next pull request will fix this code by introducing model classes for ECIES and using `objectMapper.readValue(request.getInputStream(), EciesEncryptedRequest.class)`. There are more ECIES improvements in my current code, but I don't want to keep adding changes to this already huge pull request so that we can merge it soon.